Allow setting a separate timeout for instance registration.

kawych opened the pull request:
MaxNodeProvisionTimeProvider may be now passed via AutoscalerOptions, by default the same timeout is applied for end-to-end node provisioning and node registration after the instance is created. If needed, MaxNodeProvisionTimeProvider might be extended with other timeouts with custom logic (e.g. instance provisioning excluding registration as a node).

#### What type of PR is this?

<!--
/kind feature
-->

#### What this PR does / why we need it:
This PR allows setting max node provisioning time on higher granularity.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
Directly this PR doesn't have any effect, but it adds support to override MaxNodeProvisionTimeProvider.

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

kawych commented:
/hold
I need to add unit tests specific to this functionality

BigDarkClown commented:
/assign @BigDarkClown

BigDarkClown commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -271,6 +279,13 @@ func (csr *ClusterStateRegistry) updateScaleRequests(currentTime time.Time) {
+			continue
+		} else if csr.allInstancesAreCreated(nodeGroupName) && currentTime.Add(registrationTime).Before(scaleUpRequest.ExpectedAddTime) {
+			scaleUpRequest.ExpectedAddTime = currentTime.Add(registrationTime)
+		}
```
This feels extremely hacky. IIUC we want to update the `ExpectedAddTime` once all the nodes are created to handle the separate registration time limit.

Can we just add a different variable to `scaleUpRequest` that would be initially 0 and then set once the nodes are created? IMO we should rename the `ExpectedAddTime` to `ExpectedProvisionTime` and add a new `ExpectedRegistrationTime` variable that would be dedicated to registration limits. Otherwise this becomes really unreadable.

BigDarkClown commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -300,6 +315,15 @@ func (csr *ClusterStateRegistry) updateScaleRequests(currentTime time.Time) {
 	csr.scaleDownRequests = newScaleDownRequests
 }
 
+func (csr *ClusterStateRegistry) allInstancesAreCreated(nodeGroupId string) bool {
```
Firstly, this should be called `areAllNodeCreated`. Secondly, this does not use passed `nodeGroupId`, which makes me think that it is not working as intended. Maybe with a single ongoing scale-up it works as expected, but I totally don't understand how this will behave if there are multiple ongoing scale0ups at once, especially if those would be in the same node group.

kawych replied:
Yeah, thanks for spotting this. Added filtering by nodeGroupId and extended TestScaleUpNodeRegistrationTimeout test so that reversing this change will fail an unit test.

BigDarkClown commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -1023,6 +1050,13 @@ func expectedToRegister(instance cloudprovider.Instance) bool {
 	return instance.Status != nil && instance.Status.State != cloudprovider.InstanceDeleting && instance.Status.ErrorInfo == nil
 }
 
+func provisionedTime(instance cloudprovider.Instance, currentTime time.Time) time.Time {
```
This method is used in a single place and is totally non descriptive, please inline it instead.

BigDarkClown commented on cluster-autoscaler/clusterstate/providers/max_node_provision_time_provider.go:
```diff
@@ -25,16 +26,34 @@ import (
 }
 
-// GetMaxNodeProvisionTime returns MaxNodeProvisionTime value that should be used for the given NodeGroup.
+func (p *defultMaxNodeProvisionTimeProvider) Initialize(context *context.AutoscalingContext, nodeGroupConfigProcessor nodegroupconfig.NodeGroupConfigProcessor) {
```
Please add comment

BigDarkClown commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -1246,3 +1280,16 @@ func (csr *ClusterStateRegistry) GetScaleUpFailures() map[string][]ScaleUpFailur
 }
+
+// ShouldRegisterByNow returns true i.f.f. the given node should be registered by now, according to the provided time.
+func (csr *ClusterStateRegistry) ShouldRegisterByNow(nodeGroup cloudprovider.NodeGroup, un UnregisteredNode, currentTime time.Time) (bool, error) {
```
nit: `ShouldBeProvisionedByNow`? I like this method much more than the existing approach of exposing the time limits and adding them in the call site. Could we wrap other cases in this and not expose the limits externally?

kawych replied:
Agreed, but I prefer to leave it out of this PR. Currently static_autoscaler uses MaxNodeProvisionTime in fixNodeGroupSize(), I'd prefer not to add too much scope to that. Added a TODO on MaxNodeProvisionTime and made maxNodeRegisterTime private.

BigDarkClown reviewed (commented):
Overall high level comments on this PR:

- Time limit naming is inconsistent across these changes. You are using multiple terms for nodes (nodes, intances), provisioning time (provisioning, end-to-end) or registration time (registration, instalation). Please unify the naming convention, this code is confusing to read right now.

- I understand that the maxNodeProvisionTimeProvider was added to AutoscalerOptions to enable switching it in the forked repositories. I however don't think we need to duplicate the interface and mock implementation in the `core` package - we should use the one defined in `clusterstate`. If the cyclic dependency is an issue we can move it to `clusterstate.providers`, this should resolve the issues.

- I think there are some bugs in the implementation, mostly in the `allInstancesAreCreated` method. I am not certain on how this code will behave if there are multiple ongoing scale-ups, some of which have registration issues. Has this case been tested? 

I feel like these are the most urgent comments, I'll do another round of review once those are addressed.

kawych commented:
Thanks for the review. Addressing comments, in a meantime a quick answer to your bigger comments, as it could spark some discussion
1. I can fix the consistency. Some of that is based on the existing conventions, CA already has a notion of Instance in addition to Node. I understand "node" as a K8s object, and instance as an underlying VM. Do you prefer to just consistently stick to terms "node", "provisioning" and "registration"? (regarding "provisioning", I think it's useful to explain at least once that it refers to "creation requested -> node registered" time, as my intuition would be to use this term for just VM provisioning).
2. Yeah, this is caused by circular dependencies. I'm not sure if it can be easily avoided. Now, clusterstate has to get maxNodeProvisionTimeProvider passed, this one gets AutoscalingContext in the "initialize" function and the AutoscalingContext stores ClusterState. I figured one version to untangle that would be to create a separate maxNodeProvisionTimeProvider interface for ClusterState's own use, that doesn't have Initialize function.
3. Thanks, great catch! I'll make sure to add test cases to cover that

BigDarkClown commented:
1. Yes, I think sticking to "nodes", "provisioning", "creation" and "registration" would be best. I totally agree that those need to be explained at some point, they are not that straightforward. But doing that once and then sticking to it will make the code more readable IMO.
2.  Yes, this is a bit messy at the moment. I have an idea on how we could simplify it, I'll give it a try.

kawych commented on cluster-autoscaler/clusterstate/clusterstate.go:
```diff
@@ -51,8 +51,10 @@ var (
 )
 
 type maxNodeProvisionTimeProvider interface {
```
Added a larger comment to explain this differentiation
